src/clientSideScene/sceneEntities.ts

```diff
@@ -803,53 +977,85 @@ export class SceneEntities {
-        const type = group?.userData?.type
-        const factor =
-          (sceneInfra.camControls.camera instanceof OrthographicCamera
-            ? orthoFactor
-            : perspScale(sceneInfra.camControls.camera, group)) /
-          sceneInfra._baseUnitMultiplier
-        if (type === TANGENTIAL_ARC_TO_SEGMENT) {
-          this.updateTangentialArcToSegment({
-            prevSegment: sgPaths[index - 1],
-            from: segment.from,
-            to: segment.to,
-            group: group,
-            scale: factor,
-          })
-        } else if (type === STRAIGHT_SEGMENT) {
-          this.updateStraightSegment({
-            from: segment.from,
-            to: segment.to,
-            group: group,
-            scale: factor,
-          })
-        } else if (type === PROFILE_START) {
-          group.position.set(segment.from[0], segment.from[1], 0)
-          group.scale.set(factor, factor, factor)
-        }
-      }
-      updateSegment(sketchGroup.start, 0)
-      sgPaths.forEach(updateSegment)
+      )
     })()
   }
 
+  /**
+   * Update the THREEjs sketch entities with new segment data
+   * mapping them back to the AST
+   * @param segment
+   * @param index
+   * @param varDecIndex
+   * @param modifiedAst
+   * @param orthoFactor
+   * @param sketchGroup
+   */
+  updateSegment = (
+    segment: Path | SketchGroup['start'],
+    index: number,
+    varDecIndex: number,
+    modifiedAst: Program,
+    orthoFactor: number,
+    sketchGroup: SketchGroup
+  ) => {
+    const segPathToNode = getNodePathFromSourceRange(
+      modifiedAst,
+      segment.__geoMeta.sourceRange
+    )
+    const sgPaths = sketchGroup.value
+    const originalPathToNodeStr = JSON.stringify(segPathToNode)
+    segPathToNode[1][0] = varDecIndex
+    const pathToNodeStr = JSON.stringify(segPathToNode)
+    // more hacks to hopefully be solved by proper pathToNode info in memory/sketchGroup segments
+    const group =
+      this.activeSegments[pathToNodeStr] ||
+      this.activeSegments[originalPathToNodeStr]
+    // const prevSegment = sketchGroup.slice(index - 1)[0]
+    const type = group?.userData?.type
+    const factor =
+      (sceneInfra.camControls.camera instanceof OrthographicCamera
+        ? orthoFactor
+        : perspScale(sceneInfra.camControls.camera, group)) /
+      sceneInfra._baseUnitMultiplier
+    if (type === TANGENTIAL_ARC_TO_SEGMENT) {
+      this.updateTangentialArcToSegment({
+        prevSegment: sgPaths[index - 1],
+        from: segment.from,
+        to: segment.to,
+        group: group,
+        scale: factor,
+      })
+    } else if (type === STRAIGHT_SEGMENT) {
+      this.updateStraightSegment({
+        from: segment.from,
+        to: segment.to,
+        group,
+        scale: factor,
+      })
+    } else if (type === PROFILE_START) {
+      group.position.set(segment.from[0], segment.from[1], 0)
+      group.scale.set(factor, factor, factor)
+    }
+  }
+
```

Comment:
This is an annoying diff to read but what happened here is `updateSegment` was a little helper function inside `setupSketchIdleCallbacks`, but in order to make use of it up in `setupDraftRectangle` we had to pull it out and make it a top-level method on the class.